src/livestatus.c

```diff
@@ -0,0 +1,463 @@
+
+again:
+  written = write(sockfd, request, sizeof(request));
+  if (written < 0) {
+    if (errno != EINTR) {
+      ERROR("livestatus plugin: writing to socket: %s", STRERRNO);
+      return -1;
+    }
+    goto again;
+  }
+
+  if (shutdown(sockfd, SHUT_WR) < 0) {
+    ERROR("livestatus plugin: closing socket write: %s", STRERRNO);
+    return -2;
+  }
+
+  return 0;
+} /* int ls_send_request */
+
+static value_list_t ls_collectd_init_vl(void) {
+  value_list_t vl = VALUE_LIST_INIT;
+
+  vl.values_len = 1;
+  sstrncpy(vl.plugin, LIVESTATUS_PLUGIN_NAME, sizeof(vl.plugin));
+
+  return vl;
+} /* value_list_t ls_collectd_init_vl */
+
+static int ls_collectd_dispatch_count(gauge_t value,
+                                      const char *plugin_instance) {
+  value_list_t vl = ls_collectd_init_vl();
+
+  sstrncpy(vl.type, "count", sizeof(vl.type));
+  sstrncpy(vl.plugin_instance, (char *)plugin_instance,
+           sizeof(vl.plugin_instance));
+  vl.values = &(value_t){.gauge = value};
+
+  return plugin_dispatch_values(&vl);
+} /* int ls_collectd_dispatch_count */
+
+static int ls_collectd_dispatch_counter(counter_t value,
+                                        const char *plugin_instance) {
+  value_list_t vl = ls_collectd_init_vl();
+
+  sstrncpy(vl.type, "counter", sizeof(vl.type));
+  sstrncpy(vl.plugin_instance, (char *)plugin_instance,
+           sizeof(vl.plugin_instance));
+  vl.values = &(value_t){.counter = value};
+
+  return plugin_dispatch_values(&vl);
+} /* int ls_collectd_dispatch_counter */
+
+#define L_DISPATCH(f, v, plinst)                                               \
+  if (f(v, plinst) < 0) {                                                      \
+    ERROR("livestatus plugin: fail to dispatch value for " plinst);            \
+    rc = -1;                                                                   \
+  }
+
+static int ls_collectd_dispatch(const livestatus_status_t *status) {
+  int rc = 0;
+
+  L_DISPATCH(ls_collectd_dispatch_count, status->cached_log_messages,
+             "cached_log_messages");
+  L_DISPATCH(ls_collectd_dispatch_counter, status->connections, "connections");
+  L_DISPATCH(ls_collectd_dispatch_count, status->connections_rate,
+             "connections_rate");
+  L_DISPATCH(ls_collectd_dispatch_counter, status->forks, "forks");
+  L_DISPATCH(ls_collectd_dispatch_count, status->forks_rate, "forks_rate");
+  L_DISPATCH(ls_collectd_dispatch_counter, status->host_checks, "host_checks");
+  L_DISPATCH(ls_collectd_dispatch_count, status->host_checks_rate,
+             "host_checks_rate");
+  L_DISPATCH(ls_collectd_dispatch_counter, status->livecheck_overflows,
+             "livecheck_overflows");
+  L_DISPATCH(ls_collectd_dispatch_count, status->livecheck_overflows_rate,
+             "livecheck_overflows_rate");
+  L_DISPATCH(ls_collectd_dispatch_counter, status->livechecks, "livechecks");
+  L_DISPATCH(ls_collectd_dispatch_count, status->livechecks_rate,
+             "livechecks_rate");
```

Comment:
Indeed, this seems like a good idea to drop the `_rate` metrics.
On the other hand, I do not understand why `DERIVE` would be more suitable than `COUNTER` here. Can you explain ?

Reply:
Use of `DERIVE` will calculate and provide 'rate' metric, based on collectd read interval.
So, that would be current rate, not 'average rate since launch' or 'rate for unknown period [defined in nagios daemon]'.

Reply:
collectd will also compute the rate for `COUNTER`. I think the latter is more adapted to the case at hand, as it has the advantage of not missing a value during counter reset

Reply:
The difference between `COUNTER` and `DERIVE` is that `COUNTER` tries to detect an overflow. If, in fact, the counter was reset (application got restarted, switch port went down/up), this "overflow" can result in huge and incorrect spikes. A `DERIVE` (with a minimum value of zero) will interpret a reset as a *negative* rate and dismiss the value.

With 64bit counters, overflow are very very rare and resets several orders of magnitude more likely. Hence, we recommend to use `COUNTER` only if you know exactly what you are doing. An example would be reading a 32 bit octet counter from a 1 GBit/s interface every 30 seconds: the 32 bit counters may overflow every ~34 seconds under full load.